crates/bevy_ecs/src/observer/mod.rs

```diff
@@ -0,0 +1,599 @@
+                res.assert_order(0);
+                commands.entity(obs.source()).insert(B);
+            },
+        );
+        world.observer(
+            |obs: Observer<OnRemove, A>, mut res: ResMut<R>, mut commands: Commands| {
+                res.assert_order(2);
+                commands.entity(obs.source()).remove::<B>();
+            },
+        );
+
+        world.observer(
+            |obs: Observer<OnAdd, B>, mut res: ResMut<R>, mut commands: Commands| {
+                res.assert_order(1);
+                commands.entity(obs.source()).remove::<A>();
+            },
+        );
+        world.observer(|_: Observer<OnRemove, B>, mut res: ResMut<R>| {
+            res.assert_order(3);
+        });
+
+        let entity = world.spawn(A).flush();
+        let entity = world.get_entity(entity).unwrap();
+        assert!(!entity.contains::<A>());
+        assert!(!entity.contains::<B>());
+        assert_eq!(4, world.resource::<R>().0);
+    }
+
+    #[test]
+    fn observer_multiple_listeners() {
+        let mut world = World::new();
+        world.init_resource::<R>();
+
+        world.observer(|_: Observer<OnAdd, A>, mut res: ResMut<R>| res.0 += 1);
+        world.observer(|_: Observer<OnAdd, A>, mut res: ResMut<R>| res.0 += 1);
+
+        world.spawn(A).flush();
+        assert_eq!(2, world.resource::<R>().0);
+    }
+
+    #[test]
+    fn observer_multiple_events() {
+        let mut world = World::new();
+        world.init_resource::<R>();
+        world.init_component::<A>();
+
+        world
+            .observer_builder::<OnAdd>()
+            .on_event::<OnRemove>()
+            .run(|_: Observer<_, A>, mut res: ResMut<R>| res.0 += 1);
+
+        let entity = world.spawn(A).id();
+        world.despawn(entity);
+        assert_eq!(2, world.resource::<R>().0);
+    }
+
+    #[test]
+    fn observer_multiple_components() {
+        let mut world = World::new();
+        world.init_resource::<R>();
+        world.init_component::<A>();
+        world.init_component::<B>();
+
+        world.observer(|_: Observer<OnAdd, (A, B)>, mut res: ResMut<R>| res.0 += 1);
+
+        let entity = world.spawn(A).id();
+        world.entity_mut(entity).insert(B);
+        world.flush();
+        assert_eq!(2, world.resource::<R>().0);
+    }
+
+    #[test]
+    fn observer_despawn() {
+        let mut world = World::new();
+        world.init_resource::<R>();
+
+        let observer = world
+            .observer(|_: Observer<OnAdd, A>| panic!("Observer triggered after being despawned."));
+        world.despawn(observer);
+        world.spawn(A).flush();
+    }
+
+    #[test]
+    fn observer_multiple_matches() {
+        let mut world = World::new();
+        world.init_resource::<R>();
+
+        world.observer(|_: Observer<OnAdd, (A, B)>, mut res: ResMut<R>| res.0 += 1);
+
+        world.spawn((A, B)).flush();
+        assert_eq!(1, world.resource::<R>().0);
```

Comment:
Ah, it's the `last_event_id` check. I think like there could be a comment added to the `last_event_id` to explain that this is to avoid running a single observer runner on the same event multiple times